### ISSUE TITLE
Prevent layout shifts when jumping between sections on the unified dashboard 

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#4641]
## Feature Description

As recent QA reviews have pointed out, there are some problematic side effects that come with using `useInViewSelect`, particularly when scrolling through/past multiple sections. This has the potential to cause lazy-loaded sections to be active as the user is scrolling past them when jump-scrolling to a lower section using the navigation or even landing on the page with a specific hash anchor in the URL (such as from PSI settings, or Idea Hub link).

We generally try to keep loading placeholders for skeleton component views consistent with their height with the loaded state to avoid such shifts, but this is not possible with many widgets today without forcing a specific height which could add extra space in some cases and require a scrollable widget in others.

By only activating in-view selectors when the section's (widget context ID) hash matches the hash in the URL, we can ensure only the context identified by the hash will be activated once it is in view. Alternatively, we could adjust this so that it is considered "in view" as soon as the hashes match even if the section isn't in view yet so that loading could start as soon as the hash is set. This would help reduce delay and visible loading states, particularly when the data is already cached/available. 

A proof of concept is available for review here:
https://github.com/google/site-kit-wp/pull/4634

Originally raised in https://github.com/google/site-kit-wp/issues/4410#issuecomment-1009694179

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* When navigating to a specific section on the unified dashboard (e.g. `speed`), either programmatically or by clicking on an item in the dashboard navigation, the `inView` state used by `useInViewSelect` should only be determined by a match between each dashboard section's ID  and the current URL hash
* The default inView logic for determining if an area is in view should be updated as follows:
	* The intersection observer should only consider the area below the dashboard navigation (as anything above would be covered/below)
	* The intersection should require a minimum of 2x the responsive grid gap of the area to be in view
* The state should otherwise continue to work as it does today
* This change should not affect the non-unified dashboard

#### Define `rootMargin` for the Intersection Observer in `WidgetAreaRenderer`

Within new file `assets/js/util/header.js`:

- Create and export a new function called e.g. `getHeaderHeight`.
- Extract the following block from `assets/js/util/scroll.js` as the body of `getHeaderHeight`.
  https://github.com/google/site-kit-wp/blob/2f97bc9b47cd1b11d42a9c864d8b0b00445511a5/assets/js/util/scroll.js#L36-L48
- Modify the code to be safe if `.googlesitekit-header` does not exist.
- Return `headerHeight + navigationHeight` from the function.

Within `assets/js/util/scroll.js`:

- Replace the extracted code block with the new `getHeaderHeight` function.

Within `assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js`:

- Retrieve the current `breakpoint` via `useBreakpoint`.
- Define `gridGap` according to the value of `breakpoint`. These values are double the actual grid gap, as per the AC:
  - `xlarge: 48`
  - `desktop: 48`
  - `tablet: 32`
  - `small: 32`
- Define **`rootMargin`** as follows:
  - The top should be a negative margin of the header height plus the grid gap.
  - The remaining properties should be a negative margin of the grid gap.
  - In other words, something like this:

```js
`${-(getHeaderHeight() + gridGap)}px -${gridGap}px -${gridGap}px -${gridGap}px`;
```

- It's advisable to wrap the above into a function (maybe a hook, to encapsulate the `useBreakpoint` call) for the sake of code clarity.

### Define the scrolling inView logic

Within `assets/js/googlesitekit/datastore/ui/constants.js`

- Create and export a new constant called `UI_VIEW_CONTEXT` with value `viewContext` (feel free to use a better name if you think of one). This is used as a key for `getValue` / `setValue` as you will see.

Within `assets/js/components/DashboardNavigation.js`:

- Initialise `isJumpingTo` state with the current `hash.substring( 1 )` or `undefined`.

```js
const [isJumpingTo, setIsJumpingTo] = useState(
  global.location.hash?.substring(1) || undefined
);
```

- Retrieve `setValue` from the `CORE_UI` store: `const { setValue } = useDispatch( CORE_UI )`.
- In the `handleSelect` function:
  - After the line that defines `chipID`:
    - Add a call to ```global.history.replaceState( {}, '', `#${ chipID }` )```
    - In a `setTimeout` with a short delay (say `50` ms), add a call to `setValue( UI_VIEW_CONTEXT, chipID )`.
  - The above code sets the hash when clicking on the chip, instead of setting it once the area has been scrolled to. This allows the area to begin loading as soon as the navigation is initiated so it's loaded/loading by the time it comes into view. The timeout is to make the scroll more responsive, allowing the scroll action to begin prior to the lazy loaded area rendering.
- In the `setMount` callback:
  - Replace the call to `setIsJumpingTo` with a call to `setValue( UI_VIEW_CONTEXT, isJumpingTo )`.
  - Tidy up the callback, as `isJumpingTo` can now be used in place of `chipID` and `hash` (make sure a properly prefixed hash is passed to `getContextScrollTop` though).
- In the `changeSelectedChip` function:
  - Replace the call to `global.history.replaceState` with a call to `setValue( UI_VIEW_CONTEXT, undefined )`.
- Above the second call to `changeSelectedChip`, add a call to `` global.history.replaceState( {}, '', `#${ closestID }` ) ``.
  https://github.com/google/site-kit-wp/blob/2f97bc9b47cd1b11d42a9c864d8b0b00445511a5/assets/js/components/DashboardNavigation.js#L226

Within `assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js`:

- Create a new prop `contextID` of type `string`.
- Select the `UI_VIEW_CONTEXT` value from the `CORE_UI` store:

```js
const viewContext = useSelect((select) =>
  select(CORE_UI).getValue(UI_VIEW_CONTEXT)
);
```

- Define the `value` property of `inViewState` as follows (remember `value` is set in two places):
  - If `viewContext` is truthy, `value` is the result of `viewContext === contextID`.
  - Otherwise, `value` is the same as usual: `!! intersectionEntry?.intersectionRatio`.
- Again, wrap the above into a function for code clarity/reuse.

Within `assets/js/googlesitekit/widgets/components/WidgetContextRenderer.js`:

- Pass `id` as the `contextID` prop to the `WidgetAreaRenderer`.

### Test Coverage

- Update any failing tests.

## QA Brief

* Go to the unified dashboard page.
* Click on the Speed chip
* Refresh the browser and verify that you don't see layout shifts when the page is reloaded.

## Changelog entry

* Enhance lazy loading of sections on the unified dashboard to reduce layout shifts.

**Comment by felixarntz** [comment]
@aaemnnosttv ACs LGTM!

**Comment by techanvil** [comment]
Looks like we'd better hold off the IB until https://github.com/google/site-kit-wp/issues/4488 is finalised, as it will have a significant impact on the implementation of this ticket.

**Comment by techanvil** [comment]
Hi @aaemnnosttv, having spent some time looking into this, I've come up with a potential solution that slightly diverges from that which is proposed.

It should be noted that due to changes made since your initial PoC, that approach no longer works in isolation and further changes are necessary to get the required UX to work.

I've created another [PoC](https://github.com/google/site-kit-wp/compare/poc/4410-in-view-control-mk2) to illustrate. The [first commit](https://github.com/google/site-kit-wp/commit/2ae5804904c37862bc28539184cad8f7f5318d39) in the PoC does follow the AC, in that it determines the widget lazy-load state by checking the widget's context ID against the hash:

https://github.com/google/site-kit-wp/blob/2ae5804904c37862bc28539184cad8f7f5318d39/assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js#L77-L84

This does work, to a degree. However, due to the way the context hash is updated, it results in some less than ideal UX, whereby a widget area can be scrolled significantly into the view while still remaining unloaded. In this video you can see that when navigating to Speed, scrolling up shows that the Content area lazy loads as expected, but scrolling down you can see that Monetization only loads when the Speed context is fully off the page.

![lazy-load-1](https://user-images.githubusercontent.com/18395600/149989463-e41b3668-4e13-44b6-9121-de870adc8c41.gif)

In order to get a more snappy user experience I've made a [second commit](https://github.com/google/site-kit-wp/commit/6ab996ea76597c7f5602640e33c092fabeda752f), which makes the following change (along with supporting changes):

https://github.com/google/site-kit-wp/blob/6ab996ea76597c7f5602640e33c092fabeda752f/assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js#L87-L96

In the corresponding video you can see that when navigating to Speed, the Monetization area does become active as soon as it's in view. I then reload the page and click on Content and then Monetization to highlight that the lazy loading is still working as expected.

![lazy-load-2](https://user-images.githubusercontent.com/18395600/149990655-a3ba11db-1b78-41bb-89e6-0fcb6abadabc.gif)

I know we've spoken about changing the heuristics/conditions for determining which is the "current" context, but even if we do, I still foresee the scenario where we could have widget areas with a high proportion of visibility on the screen but which are not determined to be in the current context and therefore stay unloaded, if we were to stick with the original AC.

I therefore am leaning quite in favour of the solution I've put together in the PoC. My only reservation is that it's slightly harder to follow. I have a particular gripe with this line, where I'm adding `contextHash` to the `useEffect` dependency list *purely* to trigger the effect when it changes, as it's not actually being used within the effect callback. But, I couldn't find a more elegant solution, although maybe with more time or another pair of eyes this could be improved upon.

https://github.com/google/site-kit-wp/blob/6ab996ea76597c7f5602640e33c092fabeda752f/assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js#L96

Could I ask you to take a look at the approach that I've suggested, and also bear in mind the issue that I've highlighted with following the AC as it stands, and get your opinion on whether we should go ahead with the new approach?

Thanks!

cc @felixarntz

**Comment by aaemnnosttv** [comment]
Thanks for the breakdown @techanvil.

Now that the active item in the navigation is only set once the top of the area reaches the bottom of the header (as is the common "scroll-spy" behavior), we'll need to rethink this a bit to achieve the experience we want.

At a high-level, I would say we want to control inView selects to be enabled (i.e. inView) differently depending on whether the user is "jumping" to a section (via nav) or manually scrolling.

- If jumping, the inView state for the associated area should be set by the hash match only.
	- This way the area is already loading/loaded by the time the user gets there.
	- Looking at your POC, we could have a similar "is jumping" UI state and if this is true, inView state would be determined by hash match only.
	- Any time the hash matches, the associated area should be considered inView
- If scrolling manually, we still want the hash-matching area to be inView, but we also want to consider viewport intersection so that any section with more than 5% (or maybe something absolute like 20px) on screen (whichever is greater) is also considered inView, without changing the way the active hash is set for now. (We may want to tune this a bit more but that's out of scope here I think)
	- We may also want to throttle/debounce this more than it is now to emulate the benefit we get from the "suspended" inView state we get from the jumping strategy to avoid activating several areas as a result of a quick manual scroll.
	- We'd also want to make sure that once a jump has finished that this mechanic didn't undermine the effort by triggering nearby area above which could again raise the problem this issue is intended to solve.
- If no scrolling is happening, the inView state should be set by hash-match or intersection (as above) but probably using a less-aggressive percentage to require more of the area to be visible (this is particularly important for areas above the current area rather than below, so maybe we can consider that as well). 

Regarding the inView heuristic when considering viewport intersection, we should see what works best here. As mentioned above, I think considering an area with only 1px in view would be too aggressive.

The main difference I'm describing is that multiple areas could be considered inView, which is different than the current ACs would allow, but that it shouldn't always be determined by viewport intersection, such as when jumping between.

As an aside, I think we could simplify some of the logic needed for determining which header elements are needed to calculate the header height for offsetting the intersection by introducing a new class or data attribute that we could add to all of them. That way we could 	`querySelectorAll` them with a single selector and then sum up their values as needed.

How does that sound to you @techanvil ?

Thoughts @tofumatt @felixarntz @eugene-manuilov ?

If the above sounds good, I'll update the ACs.

**Comment by eugene-manuilov** [comment]
Great work, @techanvil. I like your PoC. 

> Thoughts @tofumatt @felixarntz @eugene-manuilov ?

@aaemnnosttv, overall sounds good to me. All are good points although there are some technical details that I think we can address during IB review.

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, that broadly LGTM, but I'd question whether it's worth attempting to distinguish between scrolling manually and not scrolling. I think that would a fair amount of complexity to the scroll handling that would not be worth the gain of a marginal difference in the way intersections are calculated.

Also, given that the user would generally have to scroll to a position, during which the 5%/20px intersection would kick in and bring peripheral areas inView, if we then determine the user to have stopped scrolling, and make the intersection rules less aggressive, the areas would already be visible anyway.

Bear in mind that the only way (that I am aware of) to determine that a user is not scrolling (beyond the initial state) is to set a timeout in an onscroll handler, which means there's going to be some latency in play there too.

**Comment by aaemnnosttv** [comment]
> I'd question whether it's worth attempting to distinguish between scrolling manually and not scrolling. I think that would a fair amount of complexity to the scroll handling that would not be worth the gain of a marginal difference in the way intersections are calculated.

@techanvil the reason it's important is that we don't want to trigger every area in between the top of the page and the target of the scroll along the way. There are maybe other ways to avoid this rather than handling the programmatic scroll vs manual but this seems to be the most straightforward, although I'm open to other ideas of course.

> Bear in mind that the only way (that I am aware of) to determine that a user is not scrolling (beyond the initial state) is to set a timeout in an onscroll handler, which means there's going to be some latency in play there too.

Good point. Maybe there would need to be some compromise between the manual scrolling and idle states in the behavior as these are mostly the same. I maintain that the important change here is that programmatic scrolling to an area should only activate that one area and not the others along the way 🙂

**Comment by techanvil** [comment]
@aaemnnosttv, agreed, that's a key consideration. The way I addressed this in my PoC is, don't connect the scroll handler in `DashboardNavigation` (which is responsible for updating the hash)  when the navigation is triggered by clicking on a chip/banner, i.e. `isScrolling` is true:

https://github.com/google/site-kit-wp/blob/6234ec276f15467e747962f5c121a087a32668b2/assets/js/components/DashboardNavigation.js#L145-L150

**Comment by tofumatt** [comment]
Broadly the approach outlined by Evan here: https://github.com/google/site-kit-wp/issues/4641#issuecomment-1016266278 makes sense to me, and I'm happy with those as the ACs. Leaving any further discussion to the IB sounds good to me, like Eugene said.

I am starting to really wonder if this `useInView` approach is getting too complex though. If we can make things work very well with few glitches it sounds good, but I'm wondering if a better approach could be to look at alternative methods to prevent "too much simultaneous" requests, which is the aim of that hook and approach.

I wonder if creating a queue for all of our `API.get` calls to use, preventing more than a few requests at a time, would be better/easier-to-maintain. At this point, there's a lot of complexity to this viewport handling, such that it's a bit tough to easily grok I think.

I think these ACs are fine, but if we're in the proof-of-concept territory anyway, I _do_ wonder about not going too far in on this concept and trying an `API.get` queue instead. 🤷🏻‍♂️

**Comment by felixarntz** [comment]
@tofumatt @aaemnnosttv @techanvil Revisiting this, I'm also wondering now whether we're going a bit overboard. When I tested the unified dashboard navigation, I found the overall experience okay enough to at least not rush a fix for this - IMO we can launch without it, which allows us to take a step back and rethink the implementation. I think the severity of the layout shifts is not as critical since they at least only surface in the loading states of widgets, which realistically only are slow enough to matter when making API requests. On subsequent page loads with cached API responses, it's not as relevant.

One thing I want to call out here that was weird to me though is that, even on page loads where all API responses were cached, for some reason clicking the "Speed" pill always scrolled me in a way that the "Speed" section wasn't actually at the top of the dashboard, but somewhere in the middle (screenshot below). I would expect that if I click it super-quickly, even before the "cached API request" has run through, but it even happened for me when I waited for more than 10 seconds. It only happens if I click the "Speed" pill as very first tab on the page though, on any click afterwards it scrolls correctly. Any idea where that is coming from?

<img width="1623" alt="Screen Shot 2022-01-22 at 1 28 45 PM" src="https://user-images.githubusercontent.com/3531426/150638553-e92619af-cc47-4cdd-a7a2-e8123aaaa1c5.png">

**Comment by FlicHollis** [comment]
HI @felixarntz Reading your above comment, I will remove from this sprint and release so we can finish the IB and add to the next sprint. Let me know if you have any concerns with this, thanks!
cc @techanvil @tofumatt @aaemnnosttv

**Comment by aaemnnosttv** [comment]
> One thing I want to call out here that was weird to me though is that, even on page loads where all API responses were cached, for some reason clicking the "Speed" pill always scrolled me in a way that the "Speed" section wasn't actually at the top of the dashboard, but somewhere in the middle (screenshot below).

@felixarntz I believe this is due to the lazy loading effect of `useInViewSelect` which keeps the widget in a loading state initially (regardless of cache) and the large disparity between the module's loading and loaded states (which should be addressed regardless IMO). As it is today, if everything is cached and I click the Speed chip on a fresh page load, I end up all the way at the bottom 
![image](https://user-images.githubusercontent.com/1621608/150764237-d92f52d9-31a2-4435-b56e-11d10d82e549.png)

This does not happen when using the POC #4634

**Comment by techanvil** [comment]
Hey @FlicHollis - it seems there are some questions to resolve about the actual requirements on this ticket, before we can tackle the IB again. Maybe it should be moved back to the AC column on the board?

**Comment by FlicHollis** [comment]
Thanks @techanvil 
@felixarntz is this a priority to get in to Sprint 68? If so please could we prioritise the questions on the AC?

**Comment by aaemnnosttv** [comment]
@techanvil does the revised AC require further revision/clarification or is this sufficient to complete the IB?

**Comment by techanvil** [comment]
@aaemnnosttv I think it's sufficient to complete the IB.

We do need to wait for #4682 to be completed first as the scroll behaviour is being modified as a result of that ticket, so I have added it as a blocker.

I'll move this over to the IB column.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil – this LGTM, the only thing I'd tweak is the estimate which should probably be a little higher to leave a bit more room for additional tests (for extracted utils, etc) not only updating existing tests. Plus the changes here are not so trivial so better to err on the side of caution 😄 I'll bump it to a 15 but let me know if you feel it should be different.

IB ✅

**Comment by eugene-manuilov** [comment]
> If that's a separate issue then this is good, but I found that shift after reload weird so I'm not sure this is fixed?

Yes, that looks like a separate issue. I can't reproduce it on my end.

**Comment by wpdarren** [comment]
@aaemnnosttv @techanvil am I right in thinking that we should test Idea hub for this too? I remember this was an issue for both PSI and Idea Hub on the main and entity dashboard. Right? Just checking since the QAB doesn't mention this.

**Comment by techanvil** [comment]
Hi @wpdarren, thanks for raising this. At present this issue only tackles navigating to a section by a) loading the page with the section hash in the URL, or b) clicking on a chip to navigate to a section.

It doesn't handle the case of clicking on the links in the PSI or Idea Hub setup notifications. Notably, the link in the Idea Hub notification doesn't link directly to a section hash, but to the widget itself, which is a bit at odds with the AC.

My opinion here is that this ticket proved rather difficult to get right even without bringing those notifications into the mix (and still not QA'd yet :crossed_fingers:) and bringing the notifications into it would add complexity to the point of making it quite hard to manage. As such I think we should look at creating a separate story to address the PSI and Idea Hub notification links, in order to address this in a more iterative and manageable fashion.

Does that sound sensible to you, @aaemnnosttv?

**Comment by aaemnnosttv** [comment]
@techanvil that sounds good to me – would you please open an issue for that?

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, I have created an issue, https://github.com/google/site-kit-wp/issues/4880. I hope it's got enough info, but please let me know if you would like me to expand on what's there. Cheers

**Comment by aaemnnosttv** [comment]
## QA ❌ 

Bringing this one back to execution to resolve the error with the change to the rootMargin as raised here: https://github.com/google/site-kit-wp/issues/4869#issuecomment-1049692867

**Comment by tofumatt** [comment]
I've left a comment on the release QA ticket; thus far I haven't been able to recreate this issue after trying on several sites https://github.com/google/site-kit-wp/issues/4869#issuecomment-1050175440

I think I'd have a proposed fix, but can't verify yet. If someone has a site where they can consistently reproduce this bug, please let me know and I can provide a zip to test with 👍🏻

**Comment by wpdarren** [comment]
@tofumatt @felixarntz I am wondering if the issue I [reported here](https://github.com/google/site-kit-wp/issues/4869#issuecomment-1049692867) was to related to the issue with Unified Dashboard appearing for new sites.  I understand from Evan, this was fixed in #4865. This morning, I have created a new site, made sure to update to the latest `main` branch and I am **unable to recreate this issue**.   So, this looks to be fixed. I created a screencast here with steps and what happened for reference. 

With this said, I am wondering if this should be moved back to approved? 

https://user-images.githubusercontent.com/73545194/155693187-1c38a76b-33a0-4d96-9403-99c3099bee72.mp4

c.c. @eclarke1

**Comment by tofumatt** [comment]
I think moving it back to approval makes sense; having looked through the code (and tested it even with mocked bad data that would likely cause those JS error to happen) I can't see a path to that error. So I think it's been fixed and we can move this to approval 🙂

**Comment by tofumatt** [comment]
Oh wait, I see this is still a thing in https://github.com/google/site-kit-wp/issues/4869#issuecomment-1050707591. 😓

**Comment by asvinb** [comment]
@tofumatt PR for you to review:
https://github.com/google/site-kit-wp/pull/4892

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

Issue reported here: https://github.com/google/site-kit-wp/issues/4869#issuecomment-1050707591 is now resolved.

Found some browser specific issues which are occurring in specific scenarios. Should I create separate issues for this ?

@tofumatt 

1. Chrome – Version 98.0.4758.102
If the user navigates from 'Traffic' or 'Content' chip to directly 'Monetization' chip and refreshes the page, then it creates layout shift.

https://user-images.githubusercontent.com/94359491/155927631-98ce80fc-c6a6-4e9c-be77-71411cb55898.mp4


2. Firefox – Version 97.0.1
After navigating to other chips, when user clicks on 'Traffic chip' and refreshes the page then it creates layout shift.

https://user-images.githubusercontent.com/94359491/155926082-a63b0da8-5d47-478c-b871-fdc1922c2cb0.mp4

3. Safari – Version 15.1
If analytics is not active and content widget is not showing data then after navigating to 'Content' chip when user refreshes the page then it shows 'Traffic' chip active instead of 'Content' Chip.


https://user-images.githubusercontent.com/94359491/155929054-6c83b6e3-f68a-4306-8897-2da949890a1a.mp4


cc @wpdarren

**Comment by wpdarren** [comment]
@mohitwp good catches.  

My opinion is at this stage, we should create a new ticket but we go on what @tofumatt @techanvil feel on this.

**Comment by tofumatt** [comment]
I think a new issue is best for that; I saw those issues too and believe they're unrelated—and lower priority.

I'll let @techanvil chime in too but I think it's fine to make a new issue for that.

**Comment by techanvil** [comment]
I'd agree with @tofumatt, I think these are best to create a new issue for. I've not managed to reproduce, so unfortunately no early clues on what's causing them. Anyway, @mohitwp please do create a new issue/issues and we can investigate. Thanks.

**Comment by mohitwp** [comment]
@techanvil @tofumatt Thank you! I will create new tickets for the issues reported here https://github.com/google/site-kit-wp/issues/4641#issuecomment-1053908308

### QA Update ✅ 

Verified:

- Issue reported here: https://github.com/google/site-kit-wp/issues/4869#issuecomment-1050707591 is now resolved.

- When UFD and Service set up is enabled. Layout shift does not appear.
- Go to settings>PSI>Go to dashboard.
           **User navigate to PSI correctly.**
- Click on any chip> Load the page.
          **User navigating to correct widget.**
- Verified in browsers like Chrome, Safari, Edge and Firefox.

**Note** : When Idea hub is enabled then layout shift issue is exist. For this we have separate ticket https://github.com/google/site-kit-wp/issues/4880.
